node/npm_dependencies.ts

```diff
@@ -14,6 +15,39 @@ import { Logger } from './logger.js'
 
 const TYPESCRIPT_EXTENSIONS = new Set(['.ts', '.cts', '.mts'])
 
+/**
+ * Turns @netlify/functions into @types/netlify__functions.
+ */
+const inferDefinitelyTypedPackage = (specifier: string) => {
+  if (!specifier.startsWith('@')) return `@types/${specifier}`
+  const [scope, pkg] = specifier.split('/')
+  return `@types/${scope.replace('@', '')}__${pkg}`
+}
+
+const detectTypes = async (filePath: string): Promise<string | undefined> => {
+  try {
+    const packageJson = await findUp('package.json', { cwd: filePath })
+    if (!packageJson) return
+    const packageJsonContents = JSON.parse(await fs.readFile(packageJson, 'utf8'))
+    // this only looks at `.types` and `.typings` fields. there might also be data in `exports -> . -> types -> import/default`.
+    // we're ignoring that for now.
+    const packageJsonTypes = packageJsonContents.types ?? packageJsonContents.typings
+    if (packageJsonTypes) return join(packageJson, '..', packageJsonTypes)
+
+    const nodeModulesFolder = await findUp('node_modules', { cwd: packageJson, type: 'directory' })
+    if (!nodeModulesFolder) return
+
+    const typesPackageJson = join(
+      nodeModulesFolder,
+      inferDefinitelyTypedPackage(packageJsonContents.name),
+      'package.json',
+    )
+    const typesPackageContents = JSON.parse(await fs.readFile(typesPackageJson, 'utf8'))
+    const typesPackageTypes = typesPackageContents.types ?? typesPackageContents.typings
+    if (typesPackageContents) return join(typesPackageJson, '..', typesPackageTypes)
+  } catch {}
```

Comment:
Did that in https://github.com/netlify/edge-bundler/pull/505/commits/59d60b46931e54dbce0f08e8563a11bf0eec79c2. ESLint forced me to introduce a `safelyDetectTypes` function so that the logic isn't nested too deep.